File: python/src/dynamixel_easy_sdk/group_executor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

################################################################################
# Copyright 2025 ROBOTIS CO., LTD.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

# Author: Hyungyu Kim

from typing import List
from typing import Optional
from enum import Enum
from dataclasses import dataclass
from dynamixel_easy_sdk.dynamixel_error import *
from dynamixel_sdk import GroupBulkRead, GroupBulkWrite, GroupSyncRead, GroupSyncWrite


class CommandType(Enum):
    WRITE = 0
    READ = 1

@dataclass
class StagedCommand:
    command_type: CommandType
    id: int
    address: int
    length: int
    data: List[int]

class GroupExecutor:
    def __init__(self, connector):
        self.connector = connector
        self.port_handler = connector._port_handler
        self.packet_handler = connector._packet_handler
        self.group_bulk_write = GroupBulkWrite(self.port_handler, self.packet_handler)
        self.group_bulk_read = GroupBulkRead(self.port_handler, self.packet_handler)
        self._staged_write_commands: List[StagedCommand] = []
        self._staged_read_commands: List[StagedCommand] = []

    def addCmd(self, command: StagedCommand):
        if command.command_type == CommandType.WRITE:
            self._staged_write_commands.append(command)
        elif command.command_type == CommandType.READ:
            self._staged_read_commands.append(command)

    def clearStagedWriteCommands(self) -> None:
        self._staged_write_commands.clear()

    def clearStagedReadCommands(self) -> None:
        self._staged_read_commands.clear()

    def executeWrite(self) -> None:
        if not self._staged_write_commands:
            raise DxlRuntimeError(DxlErrorCode.EASY_SDK_COMMAND_IS_EMPTY)

        cmds = sorted(self._staged_write_commands, key=lambda c: c.id)
        ids = [cmd.id for cmd in cmds]
        if len(ids) != len(set(ids)):
            raise DxlRuntimeError(DxlErrorCode.EASY_SDK_DUPLICATE_ID)

        ref = cmds[0]
        is_sync = all(
            (c.address == ref.address and c.length == ref.length)
            for c in cmds[1:]
        )

        if is_sync:
            self._executeSyncWrite(ref.address, ref.length)
        else:
            self._executeBulkWrite()

    def _executeSyncWrite(self, address: int, length: int) -> None:
        group = GroupSyncWrite(self.port_handler, self.packet_handler, address, length)
        for cmd in self._staged_write_commands:
            if not group.addParam(cmd.id, bytes(cmd.data)):
                raise DxlRuntimeError(DxlErrorCode.EASY_SDK_ADD_PARAM_FAIL)

        dxl_comm_result = group.txPacket()
        if dxl_comm_result != DxlErrorCode.SDK_COMM_SUCCESS:
            raise DxlRuntimeError(dxl_comm_result)

    def _executeBulkWrite(self) -> None:
        self.group_bulk_write.clearParam()
        for cmd in self._staged_write_commands:
            if not self.group_bulk_write.addParam(cmd.id, cmd.address, cmd.length, bytes(cmd.data)):
                raise DxlRuntimeError(DxlErrorCode.EASY_SDK_ADD_PARAM_FAIL)

        dxl_comm_result = self.group_bulk_write.txPacket()
        if dxl_comm_result != DxlErrorCode.SDK_COMM_SUCCESS:
            raise DxlRuntimeError(dxl_comm_result)

    # -----------------------------
    # Execute Read
    # -----------------------------
    def executeRead(self) -> List[Optional[int]]:
        if not self._staged_read_commands:
            raise DxlRuntimeError(DxlErrorCode.EASY_SDK_COMMAND_IS_EMPTY)

        cmds = sorted(self._staged_read_commands, key=lambda c: c.id)
        ids = [cmd.id for cmd in cmds]
        if len(ids) != len(set(ids)):
            raise DxlRuntimeError(DxlErrorCode.EASY_SDK_DUPLICATE_ID)

        ref = cmds[0]
        is_sync = all(
            (c.address == ref.address and c.length == ref.length)
            for c in cmds[1:]
        )

        if is_sync:
            return self._executeSyncRead(ref.address, ref.length)
        else:
            return self._executeBulkRead()

    def _executeSyncRead(self, address: int, length: int) -> List[Optional[int]]:
        group = GroupSyncRead(self.port_handler, self.packet_handler, address, length)
        for cmd in self._staged_read_commands:
            if not group.addParam(cmd.id):
                raise DxlRuntimeError(DxlErrorCode.EASY_SDK_ADD_PARAM_FAIL)

        dxl_comm_result = group.txRxPacket()
        if dxl_comm_result != DxlErrorCode.SDK_COMM_SUCCESS:
            raise DxlRuntimeError(dxl_comm_result)

        results = []
        for cmd in self._staged_read_commands:
            if not group.isAvailable(cmd.id, address, length):
                results.append(None)
                continue
            value = group.getData(cmd.id, address, length)
            results.append(value)
        return results

    def _executeBulkRead(self) -> List[Optional[int]]:
        self.group_bulk_read.clearParam()
        for cmd in self._staged_read_commands:
            if not self.group_bulk_read.addParam(cmd.id, cmd.address, cmd.length):
                raise DxlRuntimeError(DxlErrorCode.EASY_SDK_ADD_PARAM_FAIL)

        dxl_comm_result = self.group_bulk_read.txRxPacket()
        if dxl_comm_result != DxlErrorCode.SDK_COMM_SUCCESS:
            raise DxlRuntimeError(dxl_comm_result)

        results = []
        for cmd in self._staged_read_commands:
            if not self.group_bulk_read.isAvailable(cmd.id, cmd.address, cmd.length):
                results.append(None)
                continue
            value = self.group_bulk_read.getData(cmd.id, cmd.address, cmd.length)
            results.append(value)
        return results

File: c++/src/dynamixel_easy_sdk/group_executor.cpp
// Copyright 2025 ROBOTIS CO., LTD.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Hyungyu Kim

#include <algorithm>

#include "dynamixel_easy_sdk/group_executor.hpp"
#include "dynamixel_easy_sdk/connector.hpp"

// Legacy Definitions usually available, but we define if missing
#ifndef COMM_SUCCESS
#define COMM_SUCCESS 0
#endif

namespace dynamixel
{

GroupExecutor::GroupExecutor(Connector * connector)
: connector_(connector),
  port_handler_(connector->getPortHandler()),
  packet_handler_(connector->getPacketHandler()),
  group_bulk_write_(port_handler_, packet_handler_),
  group_bulk_read_(port_handler_, packet_handler_)
{}

void GroupExecutor::addCmd(StagedCommand command)
{
  std::lock_guard<std::mutex> lock(mutex_);
  if (command.command_type == CommandType::WRITE) {
    staged_write_command_list_.push_back(std::move(command));
  } else if (command.command_type == CommandType::READ) {
    staged_read_command_list_.push_back(std::move(command));
  }
}

void GroupExecutor::addCmd(Result<StagedCommand, DxlError> result)
{
  if (!result.is_ok()) {
    return;
  }

  std::lock_guard<std::mutex> lock(mutex_);
  if (result.value().command_type == CommandType::WRITE) {
    staged_write_command_list_.push_back(std::move(result.value()));
  } else if (result.value().command_type == CommandType::READ) {
    staged_read_command_list_.push_back(std::move(result.value()));
  }
}

Result<void, DxlError> GroupExecutor::executeWrite()
{
  std::lock_guard<std::mutex> lock(mutex_);
  if (staged_write_command_list_.empty()) {
    return DxlError(ErrorCode::EASY_SDK_COMMAND_IS_EMPTY);
  }
  if (checkDuplicateId(staged_write_command_list_)) {
    return DxlError(ErrorCode::EASY_SDK_DUPLICATE_ID);
  }

  const auto & reference_command = staged_write_command_list_.front();
  bool is_sync = checkSync(staged_write_command_list_);

  Result<void, DxlError> result;
  if (is_sync) {
    result = executeSyncWrite(reference_command.address, reference_command.length);
  } else {
    result = executeBulkWrite();
  }
  return result;
}

Result<std::vector<Result<int32_t, DxlError>>, DxlError> GroupExecutor::executeRead()
{
  std::lock_guard<std::mutex> lock(mutex_);
  if (staged_read_command_list_.empty()) {
    return DxlError(ErrorCode::EASY_SDK_COMMAND_IS_EMPTY);
  }
  if (checkDuplicateId(staged_read_command_list_)) {
    return DxlError(ErrorCode::EASY_SDK_DUPLICATE_ID);
  }

  const auto & reference_command = staged_read_command_list_.front();
  bool is_sync = checkSync(staged_read_command_list_);

  Result<std::vector<Result<int32_t, DxlError>>, DxlError> result;
  if (is_sync) {
    result = executeSyncRead(reference_command.address, reference_command.length);
  } else {
    result = executeBulkRead();
  }
  return result;
}

Result<void, DxlError> GroupExecutor::executeSyncWrite(uint16_t address, uint16_t length)
{
  GroupSyncWrite group_sync_write(port_handler_, packet_handler_, address, length);
  for (auto & command : staged_write_command_list_) {
    Result<void, DxlError> result = processStatusRequests(command);
    if (!result.is_ok()) {
      return result.err();
    }
    if (!group_sync_write.addParam(command.id, command.data.data())) {
      return DxlError(ErrorCode::EASY_SDK_ADD_PARAM_FAIL);
    }
  }
  int dxl_comm_result = group_sync_write.txPacket();
  if (dxl_comm_result != COMM_SUCCESS) {
    return DxlError(static_cast<ErrorCode>(dxl_comm_result));
  }
  return {};
}

Result<void, DxlError> GroupExecutor::executeBulkWrite()
{
  group_bulk_write_.clearParam();

  for (auto & command : staged_write_command_list_) {
    Result<void, DxlError> result = processStatusRequests(command);
    if (!result.is_ok()) {
      return result.err();
    }
    if (!group_bulk_write_.addParam(
        command.id,
        command.address,
        command.length,
        command.data.data()))
    {
      return DxlError(ErrorCode::EASY_SDK_ADD_PARAM_FAIL);
    }
  }

  int dxl_comm_result = group_bulk_write_.txPacket();
  if (dxl_comm_result != COMM_SUCCESS) {
    return DxlError(static_cast<ErrorCode>(dxl_comm_result));
  }
  return {};
}

Result<std::vector<Result<int32_t, DxlError>>, DxlError> GroupExecutor::executeSyncRead(
  uint16_t address,
  uint16_t length)
{
  GroupSyncRead group_sync_read(port_handler_, packet_handler_, address, length);
  for (auto & command : staged_read_command_list_) {
    if (!group_sync_read.addParam(command.id)) {
      return DxlError(ErrorCode::EASY_SDK_ADD_PARAM_FAIL);
    }
  }

  int dxl_comm_result = group_sync_read.txRxPacket();
  if (dxl_comm_result != COMM_SUCCESS) {
    return DxlError(static_cast<ErrorCode>(dxl_comm_result));
  }
  std::vector<Result<int32_t, DxlError>> result_list;
  for (auto & command : staged_read_command_list_) {
    if (!group_sync_read.isAvailable(command.id, address, length)) {
      result_list.push_back(DxlError(ErrorCode::EASY_SDK_FAIL_TO_GET_DATA));
      continue;
    }

    uint32_t value = group_sync_read.getData(command.id, address, length);
    int32_t signed_value = -1;
    if (command.length == 1) {
      signed_value = static_cast<int32_t>(static_cast<int8_t>(value & 0xFF));
    } else if (command.length == 2) {
      signed_value = static_cast<int32_t>(static_cast<int16_t>(value & 0xFFFF));
    } else if (command.length == 4) {
      signed_value = static_cast<int32_t>(value);
    }

    Result<void, DxlError> result = processStatusRequests(command, signed_value);
    if (!result.is_ok()) {
      return result.err();
    }
    result_list.push_back(signed_value);
  }

  return result_list;
}

Result<std::vector<Result<int32_t, DxlError>>, DxlError> GroupExecutor::executeBulkRead()
{
  group_bulk_read_.clearParam();

  for (auto & command : staged_read_command_list_) {
    if (!group_bulk_read_.addParam(command.id, command.address, command.length)) {
      return DxlError(ErrorCode::EASY_SDK_ADD_PARAM_FAIL);
    }
  }

  int dxl_comm_result = group_bulk_read_.txRxPacket();
  if (dxl_comm_result != COMM_SUCCESS) {
    return DxlError(static_cast<ErrorCode>(dxl_comm_result));
  }

  std::vector<Result<int32_t, DxlError>> result_list;
  for (auto & command : staged_read_command_list_) {
    if (!group_bulk_read_.isAvailable(command.id, command.address, command.length)) {
      result_list.push_back(DxlError(ErrorCode::EASY_SDK_FAIL_TO_GET_DATA));
      continue;
    }

    uint32_t value = group_bulk_read_.getData(command.id, command.address, command.length);
    int32_t signed_value = -1;
    if (command.length == 1) {
      signed_value = static_cast<int32_t>(static_cast<int8_t>(value & 0xFF));
    } else if (command.length == 2) {
      signed_value = static_cast<int32_t>(static_cast<int16_t>(value & 0xFFFF));
    } else if (command.length == 4) {
      signed_value = static_cast<int32_t>(value);
    }

    Result<void, DxlError> result = processStatusRequests(command, signed_value);
    if (!result.is_ok()) {
      return result.err();
    }
    result_list.push_back(signed_value);
  }

  return result_list;
}

bool GroupExecutor::checkDuplicateId(const std::vector<StagedCommand> & list)
{
  std::vector<uint8_t> ids;
  for (const auto & cmd : list) {
    if (std::find(ids.begin(), ids.end(), cmd.id) != ids.end()) {
      return true;
    }
    ids.push_back(cmd.id);
  }
  return false;
}

bool GroupExecutor::checkSync(const std::vector<StagedCommand> & list)
{
  if (list.empty()) {return false;}
  uint16_t addr = list[0].address;
  uint16_t len = list[0].length;
  for (const auto & cmd : list) {
    if (cmd.address != addr || cmd.length != len) {
      return false;
    }
  }
  return true;
}

Result<void, DxlError> GroupExecutor::processStatusRequests(
  StagedCommand & command, int32_t read_val)
{
  if (!command.status_check_functions.empty()) {
    // Torque Enable Status Check (Virtual Logic)
    // In real implementation, we might need to read Torque Enable from cache or device.
    // Here we assume Torque is enabled if we are writing, or based on some state.
    // For now we skip actual torque check logic implementation as it requires more state management.
  }

  if (command.operating_mode_check.has_value()) {
     // Operating Mode Check (Virtual Logic)
  }

  return {};
}

} // namespace dynamixel
